config/cloudpersisters/google/datastore_persister.go

```diff
@@ -0,0 +1,216 @@
+//
+// The Datastore should further be configured with a namespace and an entity type, both of which should be passed in
+// to the `New()` function when creating a new DatastoreConfigPersister. Do not create any entity instances by hand;
+// they should only ever be created by this package.
+//
+// Once configurations are persisted, they can be viewed on the Google Cloud admin dashboard.
+package google
+
+import (
+	"io"
+	"cloud.google.com/go/datastore"
+	"golang.org/x/net/context"
+	"google.golang.org/api/option"
+	"github.com/maniksurtani/quotaservice/config"
+	"fmt"
+	"bytes"
+	"github.com/maniksurtani/quotaservice/logging"
+	"io/ioutil"
+	"time"
+	"strings"
+	"strconv"
+)
+
+
+// storedEntity stores the configuration as a serialized protobuf, and some metadata about the configuration.
+// storedEntities use a named key, of the format "version:{version_int}", to make it efficient to retrieve a
+// specific record based on version.
+type storedEntity struct {
+	Contents []byte
+	Version int32
+	Date time.Time
+	User string
+	Hash string
+}
+
+// DatastoreConfigPersister is a config persister that makes use of Google Cloud's Datastore service.
+type DatastoreConfigPersister struct {
+	projectId string
+	namespace string
+	entity string
+	client *datastore.Client
+	*config.Notifier
+	version int
+	newVersions chan int
+}
+
+func (p *DatastoreConfigPersister) PersistAndNotify(r io.Reader) error {
+	// Persist...
+	b, e := ioutil.ReadAll(r)
+	if e != nil {
+		return e
+	}
+
+	cfg, e := config.Unmarshal(bytes.NewReader(b))
+	if e != nil {
+		return e
+	}
+
+	s := &storedEntity{Contents: b,
+		Version: cfg.Version,
+		Date: time.Unix(cfg.Date, 0),
+		User: cfg.User,
+		Hash: config.HashConfig(b)}
+
+	k := datastore.NameKey(p.entity, fmt.Sprintf("version:%v", cfg.Version), nil)
+	k.Namespace = p.namespace
+
+	_, e = p.client.RunInTransaction(context.Background(), func (tx *datastore.Transaction) error {
+		existing := &storedEntity{}
+		e := tx.Get(k, existing)
+		if e != nil && e != datastore.ErrNoSuchEntity {
+			return e
+		}
+
+		if e == nil {
+			// Fail; this version already exists; ignore?
+			logging.Printf("Version %v already exists; not clobbering.", cfg.Version)
+			return nil
+		}
+
+		_, e = tx.Put(k, s)
+		if e != nil {
+			return e
+		}
+
+		return nil
+	})
+
+	if e != nil {
+		return e
+	}
+
+	// ... and notify.
+	p.Notify()
+
+	return nil
+}
+
+func (p *DatastoreConfigPersister) ConfigChangedWatcher() <-chan struct{} {
+	return p.Notifier.Watcher
+}
+
+func (p *DatastoreConfigPersister) ReadPersistedConfig() (io.Reader, error) {
+	_, s, e := p.getLatest(false)
+	if e != nil {
+		return nil, e
+	}
+
+	p.newVersions <- int(s.Version)
+
+	return bytes.NewReader(s.Contents), nil
+}
+
+func (p *DatastoreConfigPersister) getLatest(keyOnly bool) (*datastore.Key, *storedEntity, error) {
+	var entities []*storedEntity
+	q := datastore.NewQuery(p.entity).
+		Namespace(p.namespace).
+		Order("-Version").
+		Limit(1)
+
+	if keyOnly {
```

Comment:
The function `getLatest()` returns a single entity. Hence singular, `keyOnly`. The datastore API calls this `keysOnly` because it can return more than 1 entity. But I limit this with `Limit(1)`.